app/Platform/Controllers/BeatenGamesLeaderboardController.php

```diff
@@ -90,136 +85,213 @@ public function __invoke(Request $request): View
+        $selectedKind = $validatedData['filter']['kind'] ?? 'retail';
+
+        switch ($selectedKind) {
+            case 'retail':
+                $gameKindFilterOptions['retail'] = true;
+                break;
+            case 'homebrew':
+                $gameKindFilterOptions['homebrew'] = true;
+                break;
+            case 'hacks':
+                $gameKindFilterOptions['hacks'] = true;
+                break;
+            case 'all':
+                $gameKindFilterOptions = array_fill_keys($allFilterKeys, true);
+                break;
         }
 
-        return (int) DB::select(
-            (string) DB::raw('SELECT FOUND_ROWS() as count')
-        )[0]->count;
+        return [$gameKindFilterOptions, $selectedKind];
     }
 
-    private function getUserRankingData(string $username, array $gameKindFilterOptions, ?int $targetSystemId = null): array
-    {
-        $subquery = $this->buildRankingsSubquery($gameKindFilterOptions, $targetSystemId);
+    /**
+     * @return Collection<int, array|object>
+     */
+    private function getAggregatedLeaderboardData(
+        int $currentOffset,
+        array $gameKindFilterOptions,
+        ?int $targetSystemId = null,
+    ): Collection {
+        // Fetch the aggregated leaderboard.
+        $rankings = $this->buildAggregatedLeaderboardQuery($gameKindFilterOptions, $targetSystemId)->get();
+
+        // Fetch extraneous metadata without doing joins.
+        // Joins are expensive - doing this as separate queries
+        // shaves a significant amount of time from page load.
+        return $this->attachRankingRowsMetadata($rankings, $this->pageSize, $currentOffset);
+    }
 
-        /** @var string $subqueryTable */
-        $subqueryTable = DB::raw("({$subquery->toSql()}) as b");
+    private function getIncludedTypes(array $gameKindFilterOptions = []): array
+    {
+        $includedTypes = [];
 
-        $result = DB::table($subqueryTable)
-            ->mergeBindings($subquery)
-            ->where('User', $username)
-            ->get();
+        if ($gameKindFilterOptions['retail']) {
+            $includedTypes[] = RankingType::GamesBeatenHardcoreRetail;
+        }
+        if ($gameKindFilterOptions['hacks']) {
+            $includedTypes[] = RankingType::GamesBeatenHardcoreHacks;
+        }
+        if ($gameKindFilterOptions['homebrew']) {
+            $includedTypes[] = RankingType::GamesBeatenHardcoreHomebrew;
+        }
+        if ($gameKindFilterOptions['unlicensed']) {
+            $includedTypes[] = RankingType::GamesBeatenHardcoreUnlicensed;
+        }
+        if ($gameKindFilterOptions['prototypes']) {
+            $includedTypes[] = RankingType::GamesBeatenHardcorePrototypes;
+        }
+        if ($gameKindFilterOptions['demos']) {
+            $includedTypes[] = RankingType::GamesBeatenHardcoreDemos;
+        }
 
-        $userRankingData = $result->isEmpty() ? null : $result->get(0);
+        return $includedTypes;
```

Comment:
is there still some way to select multiple types?

Reply:
Yes. The "All" radio button selects all types.